File: server/structs/mqtt.go
package structs

type MqttPayload struct {
	MsgID     string `json:"msgid"`
	Timestamp int64  `json:"timestamp"`
	Type      string `json:"type"`
	Action    string `json:"action"`
	// Add optional Data struct that can be anything
	Data interface{} `json:"data"`
}

type MqttTempatureData struct {
	CurrHotbedTemp   int `json:"curr_hotbed_temp"`
	CurrNozzleTemp   int `json:"curr_nozzle_temp"`
	TargetHotbedTemp int `json:"target_hotbed_temp"`
	TargetNozzleTemp int `json:"target_nozzle_temp"`
}

type MqttResponse struct {
	Type      string      `json:"type"`
	Action    string      `json:"action"`
	MsgID     string      `json:"msgid"`
	State     string      `json:"state"`
	Timestamp int64       `json:"timestamp"`
	Code      int         `json:"code"`
	Msg       string      `json:"msg"`
	Data      interface{} `json:"data"`
}

File: kobra-client/src/interfaces/mqtt.ts
export interface MqttResponse {
  type: string;
  action: string;
  msgID: string;
  state: string;
  timestamp: number;
  code: number;
  msg: string;
  data: any;
}

interface MqttTempatureData {
  curr_hotbed_temp: number;
  curr_nozzle_temp: number;
  target_hotbed_temp: number;
  target_nozzle_temp: number;
}

interface MqttPrintUpdateData {
  taskid: string;
  filename: string;
  print_time: number;
  progress: number;
  supplies_usage: number;
  total_layers: number;
  curr_layer: number;
  remain_time: number;
}

// Temperature interface that extends the MqttResponse interface
export interface Temperature extends MqttResponse {
  data: MqttTempatureData;
}

export interface PrintUpdate extends MqttResponse {
  data: MqttPrintUpdateData;
}
